### ISSUE TITLE
Feature request: Allow declaring main language of app to VoiceOver and TalkBack

### ISSUE THREAD

**Issue by Leffe108** [flutter/flutter#99600]
## Use case
Allow declaring the main language of an app to assistive technology such as TalkBack and VoiceOver. 

When an app correctly declare its language, the voice engine can select an appropriate voice for the content. This is important because it is common that screen readers have separate voice engines for each language. Using eg. the English language engine to read Swedish text makes it harder for the user to hear and understand what is being said. In the linked issues below there are video recordings where VoiceOver uses the wrong voice as well as sample app code to reproduce the issue.

### Supporting sources
This is a top-level recommendation in WCAG accessibility guidelines. 
* [WCAG - Language of page - level A](https://www.w3.org/WAI/WCAG21/quickref/?versions=2.0&showtechniques=312#qr-meaning-doc-lang-id)
* [WCAG also covers setting language in their development tips](https://www.w3.org/WAI/tips/developing/#identify-page-language-and-language-changes)

## Proposal
Make `MaterialApp.locale` propagate to assitive technology similar to how `<html lang="sv-SE">`  sets the language in HTML.

Alternatively, provide a widget similar to `Localizations.override` that does communicate locale to TalkBack and VoiceOver. A such widget could be used by Flutter app to wrap all their content and ma also be used to satisfy the [WCAG level AA recommendation to indicate change in language on a page](https://www.w3.org/WAI/WCAG21/Understanding/language-of-parts.html). 

## Related issues
* https://github.com/flutter/flutter/issues/99018 (code sample + video)
* https://github.com/flutter/flutter/issues/98948 (code sample + video)
* https://github.com/flutter/flutter/issues/98835 (provides 2 use cases)

**Comment by mtropper** [comment]
Any updates on the status of this feature?

**Comment by jelen-HC** [comment]
Has there been any progress on this issue? This is quite an important feature for those requiring assistive technologies and as of current there are quite a bit of short comings with how flutter handles the switching of language/locale with both Voiceover and Talkback.

**Comment by Leffe108** [comment]
It might help to add an emoji reaction to OP post in this issue and related https://github.com/flutter/flutter/issues/98948 issue. It has in the past been told that it is used by the flutter team to sort importance of issues.

**Comment by LHaanappel** [comment]
This is a pretty important feature which is holding our app back from obtaining a A status according to the accessibility audits we have done. Unless this feature is implemented Flutter apps will not be able to comply to the The European Accessibility Act which we need to comply to by 2025.

**Comment by chunhtai** [comment]
Is this for web only? If so, yes I think this is something we should do.

For mobile, I don't think there is an API to override VoiceOver language entirely. We would need to manually wrap every string with LocaleString attribute. but I am not sure if that is something we should do

**Comment by chunhtai** [comment]
> This is a pretty important feature which is holding our app back from obtaining a A status according to the accessibility audits we have done. Unless this feature is implemented Flutter apps will not be able to comply to the The European Accessibility Act which we need to comply to by 2025.

Can you explain more detail how the app fail the accessibility standard? The MaterialApp locale is coming from the OS, which should be the same as the voiceover language.

**Comment by LHaanappel** [comment]
> > This is a pretty important feature which is holding our app back from obtaining a A status according to the accessibility audits we have done. Unless this feature is implemented Flutter apps will not be able to comply to the The European Accessibility Act which we need to comply to by 2025.
> 
> Can you explain more detail how the app fail the accessibility standard? The MaterialApp locale is coming from the OS, which should be the same as the screen reader's language.

According to the accessibility audit we executed for our app, we failed WCAG Guideline 3.1.1. This guideline describes that the default human language of each webpage can be programmatically determined. The reason that was given to us why we failed the criterium was:

> The application cannot automatically detect the language, which causes the texts to be spoken
> in the system language. This means that there is no code to tell TalkBack that the
> texts must be pronounced in the language of the application. The default language of the
> application must be able to be determined by auxiliary software, this must be independent of the
> language setting of the device.

The language of our app is Dutch (supportedLocales is set to Dutch as well), but when the system language is set to English, the content of the app is spoken by the screenreader/VoiceOver as if it was English text.

Putting every text in our app in a TextSpan with a specified locale is not really a viable solution for us.

**Comment by Leffe108** [comment]
> > This is a pretty important feature which is holding our app back from obtaining a A status according to the accessibility audits we have done. Unless this feature is implemented Flutter apps will not be able to comply to the The European Accessibility Act which we need to comply to by 2025.
> 
> Can you explain more detail how the app fail the accessibility standard? The MaterialApp locale is coming from the OS, which should be the same as the screen reader's language.

# Accessibility standard
In EU there are two accessibility acts. One is already in effect and that one affects the public sector and companies that deliver custom built software to public sector. To harmonize the requirements across EU countries, there exist a standard EN301549. This standard builds upon WCAG. Initially I think 2.0, but there has been an update to include WCAG 2.1. (basically, first EU make an Act that is then implemented in each countries law (slightly differently) and then a harmonized standard provide guidance for companies to comply to the act in a uniform way) 

All information I found points towards that the harmonized standard for the new Act that becomes effective in 2025 will also build upon WCAG 2.1 requirements. The new act main purpose from my understanding is to include more sectors and types of apps to comply to accessibility standards including e-commerce and public transport websites, apps and also physical interactive devices.

Reading EN301549 - for web it require compliance with both document-level language declaration and part-level language declaration (section 9.3). For software, (section 11), the requirement appear to only be on top level, and it also include texts that if the software comply with the system advertised language, the software doesn't need to declare its language to assistive tech. 

[See section 11, page 67](https://www.etsi.org/deliver/etsi_en/301500_301599/301549/02.01.02_60/en_301549v020102p.pdf)
![screenshot of section 11.3 on page 67](https://github.com/flutter/flutter/assets/4630670/5f376d45-8ab6-4003-bfb7-6022c3fbb534)

# Material locale etc.
@chunhtai You are right, that if the app does present the text in the locale of the system, it seems that that will comply with the EN301549 standard. However, that assumes that the app either uses restricted access to only the countries that use the languages of the app, or it includes real-time translation of texts to support all languages of the globe. Otherwise, you will end up with the situation that some users have system language set to eg. Swedish but uses an app that does not support Swedish, so the localization layer will resort to displaying the texts in English and this is currently not communicated by Flutter to assistive tech. So assistive tech will use the Swedish voice to try to read out English text which will at best sound funny or at worst fail to make the information audible. 

Where i work we have an app that is translated to a few European languages including English, but it happens that the app is being used in countries for which we have not yet localized the content where they end up being served the English locale.

It might be possible for some apps to clearly declare which locales it support and not be "available" for other languages and comply by using the read system language method. But it would be helpful if the framework provided support to comply with the standard via the other way - by advertising the resolved translation/locale to the assistive tech. 

# I am not a lawyer
So, this is my interpretation of various documents, and attending a seminar by AccessibleEU, but I am not a lawyer so you may want to do your homework as well, and/or kindly point out if I misunderstood something.

**Comment by chunhtai** [comment]
yup, this sounds reasonable. This will be a non trivial change though. We may need a hook to set accessibility locale form the localization.

**Comment by proohit** [comment]
Is there any news on this? The solution using `TextSpan` via `Text.rich(TextSpan(text: ..., locale: const Locale('de', 'DE')))` works but isn't practicable in that scale.  We need that, because most of our texts are in a specific language.

**Comment by vgtle** [comment]
This is important for anyone in the EU using Flutter.  This not only applies to the `Text` Widget, but e.g. the `hintText` parameter of the `TextField` or the Semantics widget (`label`parameter) will always be in the wrong language. There are probably a lot more widgets affected by this.

Can we adjust the Text Widget implementation accordingly to pass the `locale` parameter or alternative the default locale to the `TextSpan` (e.g. `Text` Widget line 691 or 711):

``` dart
 result = RichText(
        textAlign: textAlign ?? defaultTextStyle.textAlign ?? TextAlign.start,
        textDirection: textDirection, // RichText uses Directionality.of to obtain a default if this is null.
        locale: locale, // RichText uses Localizations.localeOf to obtain a default if this is null
        softWrap: softWrap ?? defaultTextStyle.softWrap,
        overflow: overflow ?? effectiveTextStyle?.overflow ?? defaultTextStyle.overflow,
        textScaler: textScaler,
        maxLines: maxLines ?? defaultTextStyle.maxLines,
        strutStyle: strutStyle,
        textWidthBasis: textWidthBasis ?? defaultTextStyle.textWidthBasis,
        textHeightBehavior: textHeightBehavior ?? defaultTextStyle.textHeightBehavior ?? DefaultTextHeightBehavior.maybeOf(context),
        selectionColor: selectionColor ?? DefaultSelectionStyle.of(context).selectionColor ?? DefaultSelectionStyle.defaultColor,
        text: TextSpan(
          /// -----------------------------
          locale: locale ?? Localizations.localOf(context), // <- This line
          /// -----------------------------
          style: effectiveTextStyle,
          text: data,
          children: textSpan != null ? <InlineSpan>[textSpan!] : null,
        ),
      );
```

You probably know better about any side effects this could have, but this seems to be a pretty simple solution for a problem that will affect the whole app.

**Comment by chunhtai** [comment]
@vgtle 
That will probably be fine, but I think If we want to set a global locale I would rather not relying it on propagating down using default value of Textspan.locale because in most platform there is an API to set it in top level.

For iOS, you can use UIApplication.accessibilityLanguage
For web, you can set the lang attribute.
For android, there seems to be none, but we can manually wrap localespan in the engine embedding

**Comment by lColinDl** [comment]
The workaround specifying the `TextSpan.locale` works in a lot of cases, but there are also APIs where such a workaround does not exist.
Some examples that my team encountered are `CustomSemanticsAction` or `SemanticsService.announce`.

**Comment by chunhtai** [comment]
Hey @lColinDl 
That is a good point, we should support locale for those API. Can you create a new issue for those?

**Comment by JJdeGroot** [comment]
@chunhtai for Android there is a new API:

Guidance:

https://developer.android.com/guide/topics/resources/app-languages
> Additional APIs: These public APIs, such as the [setApplicationLocales()](https://developer.android.com/reference/android/app/LocaleManager#setApplicationLocales(android.os.LocaleList)) and [getApplicationLocales()](https://developer.android.com/reference/android/app/LocaleManager#getApplicationLocales()) methods in [LocaleManager](https://developer.android.com/reference/android/app/LocaleManager), let apps set a different language from the system language at runtime.

Docs:

[`public void setApplicationLocales(LocaleList locales))`](https://developer.android.com/reference/android/app/LocaleManager#setApplicationLocales(android.os.LocaleList))

> Sets the UI locales for the calling app.
>
> Pass a [LocaleList#getEmptyLocaleList()](https://developer.android.com/reference/android/os/LocaleList#getEmptyLocaleList()) to reset to the system locale.
>
> Note: Changes to app locales will result in a configuration change (and potentially an Activity lifecycle event) being applied to the calling application. For more information, see the [section on handling configuration changes](https://developer.android.com/guide/topics/resources/runtime-changes). The set locales are persisted; they are backed up if the user has enabled Backup & Restore.
>
> Note: Users' locale preferences are passed to applications by creating a union of any app-specific locales and system locales, with the app-specific locales appearing first. Language resources are then chosen per usual (as described in the [section on locale resolution](https://developer.android.com/guide/topics/resources/multilingual-support)).

Sample:

```kotlin
val locale = LocaleListCompat.forLanguageTags("nl-NL")
AppCompatDelegate.setApplicationLocales(locale)
```

**Comment by chunhtai** [comment]
web also have ability denote language in a part of page
and the WCAG also required the language attribute to be set.

It seems like we need to API that ties to the localization widget.
1. a global locale.
2. denote subtree locale.

**Comment by chunhtai** [comment]
raise to p1 since there is currently no way to accurately set lang attribute in web based on framework locale

**Comment by chunhtai** [comment]
web version https://github.com/flutter/flutter/issues/98948

**Comment by chunhtai** [comment]
For Android this correctly override the application locales
```java
import androidx.appcompat.app.AppCompatDelegate;
import androidx.core.os.LocaleListCompat;

        LocaleListCompat appLocale =
                LocaleListCompat.forLanguageTags("xx-XX");
        AppCompatDelegate.setApplicationLocales(appLocale);
```

**Comment by chunhtai** [comment]
done for mobiles, just web part left
